subcommand/tls-init/command.go

```diff
@@ -0,0 +1,368 @@
+			return 1
+		}
+	}
+
+	var cancel context.CancelFunc
+	c.ctx, cancel = context.WithCancel(context.Background())
+	defer cancel()
+
+	// Verify if CA cert and key exist as Kubernetes secrets if they are not provided as file.
+	if c.caFile == "" && c.keyFile == "" {
+		c.caCertSecret, err = c.clientset.CoreV1().Secrets(c.k8sNamespace).Get(c.ctx, fmt.Sprintf("%s-ca-cert", c.namePrefix), metav1.GetOptions{})
+		if err != nil && !k8serrors.IsNotFound(err) {
+			c.log.Error("error reading secret from kubernetes client", "err", err)
+			return 1
+		} else if err != nil {
+			c.caCertSecret = nil
+		}
+		c.caKeySecret, err = c.clientset.CoreV1().Secrets(c.k8sNamespace).Get(c.ctx, fmt.Sprintf("%s-ca-key", c.namePrefix), metav1.GetOptions{})
+		if err != nil && !k8serrors.IsNotFound(err) {
+			c.log.Error("error reading secret from kubernetes client", "err", err)
+			return 1
+		} else if err != nil {
+			c.caKeySecret = nil
+		}
+	}
+
+	// Only create a CA certificate/key pair if it doesn't exist or hasn't been provided
+	if !c.existingCA() {
+		c.log.Info("creating CA certificate and key as it does not exists")
+		_, pk, ca, _, err = cert.GenerateCA("Consul Agent CA")
+		if err != nil {
+			c.log.Error("error generating Consul Agent CA certificate and private key", "err", err)
+			return 1
+		}
+
+		c.log.Info("saving ca certificate")
+		_, err = c.clientset.CoreV1().Secrets(c.k8sNamespace).Create(c.ctx, &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      fmt.Sprintf("%s-ca-cert", c.namePrefix),
+				Namespace: c.k8sNamespace,
+			},
+			Data: map[string][]byte{
+				corev1.TLSCertKey: []byte(ca),
+			},
+			Type: corev1.SecretTypeOpaque,
+		}, metav1.CreateOptions{})
+
+		if err != nil {
+			c.log.Error("error saving CA certificate secret to kubernetes", "err", err)
+			return 1
+		}
+		c.log.Info("saving ca private key")
+		_, err = c.clientset.CoreV1().Secrets(c.k8sNamespace).Create(c.ctx, &corev1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      fmt.Sprintf("%s-ca-key", c.namePrefix),
+				Namespace: c.k8sNamespace,
+			},
+			Data: map[string][]byte{
+				corev1.TLSPrivateKeyKey: []byte(pk),
+			},
+			Type: corev1.SecretTypeOpaque,
+		}, metav1.CreateOptions{})
+
+		if err != nil {
+			c.log.Error("error saving CA private key secret to kubernetes", "err", err)
+			return 1
+		}
+		c.log.Info("successfully saved ca and private key")
+	}
+
+	var hosts []string
+	var name string
+	var caBytes, keyBytes []byte
+
+	if c.caFile != "" && c.keyFile != "" {
+		c.log.Info("reading CA certificate from provided file")
+		caBytes, err = ioutil.ReadFile(c.caFile)
+		if err != nil {
+			c.log.Error("error reading provided CA file", "err", err)
+			return 1
+		}
+		c.log.Info("reading CA private key from provided file")
+		keyBytes, err = ioutil.ReadFile(c.keyFile)
+		if err != nil {
+			c.log.Error("error reading provided private key file", "err", err)
+			return 1
+		}
+	} else {
+		if c.caCertSecret == nil {
```

Comment:
We do read it on line 107, but if it doesn't exist, we explicitly set it to `nil` on line 112. The reason we do that, is because the API server does not return a `nil` if the secret isn't found. In this case, the secret will be `nil` if we have to create the CA certificate which is the situation we will be in during a bootstrap.

Reply:
So I cleaned this up. let me know if it looks better in the current iteration.

Reply:
Looks great!